### PR TITLE
[feat]: brand new enums

### PR DESCRIPTION
**Pull request by 0x009922**
In short:

```ts
type Option<T> = Enumerate<{
  None: [],
  Some: [T]
}>

function display(maybe: Option<string>): string {
  if (maybe.tag === 'None') return '...'
  if (maybe.tag === 'Some') return `Some(${maybe.content.trim()})`

  // TypeScript knows that conditions are exhaustive (works with `switch` as well),
  // and that the function returns `string` anyway.
}

function notExhaustiveDisplay<T>(maybe: Option<T>): string {
  if (maybe.tag === 'None') return '...'

  // Whoops! Function declares `string` return type, but TypeScript detects
  // that not all code branches return `string` - compilation error
}

display(variant('Some', 'foo')) // `variant(...)` infers the types and has IDE hints

// everything below are errors
display(variant('Foo'))
display(variant('Some'))
display(variant('None', 'bar'))
display(variant('Some', false))
```

For what & why please refer to the updated CHANGELOGs and READMEs.

### CODE REVIEW

File: packages/core/CHANGELOG.md [hunk 1]
```diff
@@ -1,5 +1,17 @@
 # @scale-codec/core
 
+## 2.0.0
+
+### Major Changes
+
+- 7e81bbc: Follows the major update of the re-exported `@scale-codec/enum` lib.
+- 7e81bbc: **refactor**: "void" type is renamed to "unit", i.e. `encodeVoid` -> `encodeUnit`, `decodeVoid` -> `decodeUnit`
```
Comment (outoftardis) [inline packages/core/CHANGELOG.md:8]:
```suggestion
- 7e81bbc: **refactor**: "void" types are renamed to "unit", i.e. `encodeVoid` -> `encodeUnit`, `decodeVoid` -> `decodeUnit`
```

File: packages/core/CHANGELOG.md [hunk 2]
```diff
@@ -1,5 +1,17 @@
 # @scale-codec/core
 
+## 2.0.0
+
+### Major Changes
+
+- 7e81bbc: Follows the major update of the re-exported `@scale-codec/enum` lib.
```
Comment (outoftardis) [inline packages/core/CHANGELOG.md:7]:
```suggestion
- 7e81bbc: Follow the major update of the re-exported `@scale-codec/enum` lib.
```

I'd also suggest saying which update it was

File: packages/definition-compiler/CHANGELOG.md [hunk 3]
```diff
@@ -1,5 +1,18 @@
 # @scale-codec/definition-compiler
 
+## 3.0.0
+
+### Minor Changes
+
+- 7e81bbc: **refactor**: generate imports/exports each on a new line
```
Comment (outoftardis) [inline packages/definition-compiler/CHANGELOG.md:7]:
```suggestion
- 7e81bbc: **refactor**: generate each import/export on a new line
```

File: packages/definition-compiler/CHANGELOG.md [hunk 4]
```diff
@@ -1,5 +1,18 @@
...
+- 7e81bbc: **refactor**: update generation of enums due to the major update of their design
```
Comment (outoftardis) [inline packages/definition-compiler/CHANGELOG.md:8]:
```suggestion
- 7e81bbc: **refactor**: update enum generation following the major update in enum design
```

File: packages/definition-runtime/CHANGELOG.md [hunk 5]
```diff
@@ -1,5 +1,21 @@
 # @scale-codec/definition-runtime
 
+## 3.0.0
+
+### Major Changes
+
+- 7e81bbc: **refactor**: update due to the major update of the core library
```
Comment (outoftardis) [inline packages/definition-runtime/CHANGELOG.md:7]: you need to say what you are updating

File: packages/enum/CHANGELOG.md [hunk 6]
```diff
@@ -1,5 +1,118 @@
 # @scale-codec/enum
 
+## 2.0.0
+
+### Major Changes
+
+- 7e81bbc: **refactor**: introduce a complete new design of tagged unions
```
Comment (outoftardis) [inline packages/enum/CHANGELOG.md:7]:
```suggestion
- 7e81bbc: **refactor**: introduce a completely new design of tagged unions
```

```suggestion
- 7e81bbc: **refactor**: introduce a new design of tagged unions
```

File: packages/enum/CHANGELOG.md [hunk 7]
```diff
@@ -1,5 +1,118 @@
...
+  ```ts
+  import { variant } from '@scale-codec/enum'
+
+  let msg: Message = variant('Ping')
+  msg = variant('Text', 'foo')
+  msg = variant('Coords', [4, 2])
+  msg = variant('User', { name: 'bar' })
+  ```
+
+  **Variant instance** has `tag` and `content` fields to work with:
```
Comment (outoftardis) [inline packages/enum/CHANGELOG.md:59]:
```suggestion
  Each **variant instance** has `tag` and `content` fields to work with:
```

File: packages/enum/CHANGELOG.md [hunk 8]
```diff
@@ -1,5 +1,118 @@
...
+
+  ```ts
+  declare const msg: MessageVariant
+
+  if (msg.tag === 'Text') {
+    console.log(msg.content.trim())
+  }
+  ```
+  
+  There are other changes as well, but that's all for major ones.
```
Comment (outoftardis) [inline packages/enum/CHANGELOG.md:69]:
```suggestion
  There are other changes as well, but that's all for the major ones.
```

File: packages/enum/CHANGELOG.md [hunk 9]
```diff
@@ -1,5 +1,118 @@
...
+
+
+  #### Why the change was made
+
+  The goal is to be consistent with [TypeScript support of discriminated unions](https://www.typescriptlang.org/docs/handbook/typescript-in-5-minutes-func.html#discriminated-unions). The consequence is that TypeScript now able to narrow types and to check for exhaustiveness in code that works with enums. In other words, **new enums are type-stronger**.
```
Comment (outoftardis) [inline packages/enum/CHANGELOG.md:74]:
```suggestion
  The goal is to be consistent with [TypeScript support of discriminated unions](https://www.typescriptlang.org/docs/handbook/typescript-in-5-minutes-func.html#discriminated-unions). As a consequence, TypeScript can now narrow types and check for exhaustiveness in code that works with enums. In other words, **new enums are type-stronger**.
```

File: packages/enum/README.md [hunk 10]
```diff
@@ -1,49 +1,47 @@
...
+Small tagged-union library for TypeScript.
 
-[Documentation](https://soramitsu.github.io/scale-codec-js-library/guide/enum)
+[//]: # 'TODO'
+[//]: # '[Documentation](https://soramitsu.github.io/scale-codec-js-library/guide/enum)'
+
+## Features
+
+- Follows TypeScript's [discriminated union pattern](https://www.typescriptlang.org/docs/handbook/typescript-in-5-minutes-func.html#discriminated-unions), which leads to **type narrowing** and **exhaustiveness check** support.
+- Type-safe variants creation with type inference support
```
Comment (outoftardis) [inline packages/enum/README.md:11]:
```suggestion
- **Type narrowing** and **exhaustiveness check** (due to the library following TypeScript's [discriminated union pattern](https://www.typescriptlang.org/docs/handbook/typescript-in-5-minutes-func.html#discriminated-unions))
- Type-safe variants creation with type inference
```

File: packages/definition-runtime/CHANGELOG.md (hunk 5 above)
Comment (0x009922) [reply packages/definition-runtime/CHANGELOG.md]: Said

File: packages/core/CHANGELOG.md (hunk 2 above)
Comment (0x009922) [reply packages/core/CHANGELOG.md]: Said

File: packages/definition-runtime/CHANGELOG.md [hunk 11]
```diff
@@ -1,5 +1,22 @@
 # @scale-codec/definition-runtime
 
+## 3.0.0
+
+### Major Changes
+
+- 7e81bbc: **refactor**: update due to the major update of the core library and enum design. Update `createEnumCodec` function signature and `EnumCodecAndFactory` type.
+- 7e81bbc: **refactor**: remove `Opaque` usage from codec factories, but still provide a type argument to specify an opaque type returned by the factory
```
Comment (outoftardis) [inline packages/definition-runtime/CHANGELOG.md:8]:
```suggestion
- 7e81bbc: **refactor**: remove `Opaque` usage from codec factories but still provide a type argument to specify an opaque type returned by the factory
```

**Review by outoftardis — APPROVED** [review]

**Review by stefashkaa — APPROVED** [review]

**Pull request merged**